### PR TITLE
[OpenVINO-EP] Updating OpenVINO EP to 2020.4

### PR DESCRIPTION
**Pull request by suryasidd**
**Description**: Updating OpenVINO EP to support OpenVINO version 2020.4
- Added backward compatibility with 2020.2 and 2020.3 versions.
- Added support for more operators and models

**Motivation and Context**
- Allows users to use OpenVINO 2020.4 with OpenVINO EP

### CODE REVIEW

**Comment by HectorSVC** [comment]
There's a test failure (result diff), please take a look.
2020-08-18T07:09:50.0264924Z 1: [ RUN      ] ModelTests/ModelTest.Run/719
2020-08-18T07:09:50.0313166Z 1: Warning: ImageScaler was a removed experimental ops. In the future, we may directly reject this operator. Please update your model as soon as possible.
2020-08-18T07:09:50.0314320Z 1: Warning: Crop was a removed experimental ops. In the future, we may directly reject this operator. Please update your model as soon as possible.
2020-08-18T07:09:50.0315301Z 1: Warning: Crop was a removed experimental ops. In the future, we may directly reject this operator. Please update your model as soon as possible.
2020-08-18T07:09:50.0316198Z 1: Warning: Crop was a removed experimental ops. In the future, we may directly reject this operator. Please update your model as soon as possible.
2020-08-18T07:09:50.0317086Z 1: Warning: Crop was a removed experimental ops. In the future, we may directly reject this operator. Please update your model as soon as possible.
2020-08-18T07:09:50.0321897Z 1: Warning: Crop was a removed experimental ops. In the future, we may directly reject this operator. Please update your model as soon as possible.
2020-08-18T07:09:50.0322998Z 1: Warning: Crop was a removed experimental ops. In the future, we may directly reject this operator. Please update your model as soon as possible.
2020-08-18T07:09:50.0323965Z 1: Warning: Crop was a removed experimental ops. In the future, we may directly reject this operator. Please update your model as soon as possible.
2020-08-18T07:09:50.0324961Z 1: Warning: Affine was a removed experimental ops. In the future, we may directly reject this operator. Please update your model as soon as possible.
2020-08-18T07:11:22.4653749Z 1: /onnxruntime_src/onnxruntime/test/providers/cpu/model_tests.cc:556: Failure
2020-08-18T07:11:22.4655738Z 1: Expected equality of these values:
2020-08-18T07:11:22.4657791Z 1:   COMPARE_RESULT::SUCCESS
2020-08-18T07:11:22.4659587Z 1:     Which is: 4-byte object <00-00 00-00>
2020-08-18T07:11:22.4660015Z 1:   ret.first
2020-08-18T07:11:22.4660586Z 1:     Which is: 4-byte object <01-00 00-00>
2020-08-18T07:11:22.4661154Z 1: expected 132 (43040000), got 142.494 (430e7e7a), diff: 10.494, tol=10.0013 idx=851956. 1 of 1555200 differ
2020-08-18T07:11:22.5404617Z 1: [  FAILED  ] ModelTests/ModelTest.Run/719, where GetParam() = "openvino_../models/opset7/coreml_FNS-Candy_ImageNet/winmlperf_coreml_FNS-Candy_prerelease.onnx" (92514 ms)

File: onnxruntime/core/providers/openvino/ov_versions/capabilities.h [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+namespace onnxruntime {
+namespace openvino_ep {
+
+#if (defined OPENVINO_2020_2) || (defined OPENVINO_2020_3)
+std::vector<std::unique_ptr<ComputeCapability>>
+GetCapability_2020_2(const onnxruntime::GraphViewer& graph_viewer, const std::string device_id);
+
+#elif defined OPENVINO_2020_4
+std::vector<std::unique_ptr<ComputeCapability>>
+GetCapability_2020_4(const onnxruntime::GraphViewer& graph_viewer, const std::string device_id);
```
Comment (HectorSVC) [inline onnxruntime/core/providers/openvino/ov_versions/capabilities.h:14]: Are we going to always copy the same code and make some modification for new version? I think there are quite a lot of share code. What are you going to do if bugs found for the shared code? Should we think about a way for code sharing. I suppose there would be less limitation/rules for newer version. Is it possible to create base class with basic check/validations for newer OpenVino version, the older one inherit from that and extend the check/validations.

File: tools/ci_build/github/linux/docker/scripts/install_openvino.sh [hunk 2]
```diff
@@ -7,13 +7,14 @@ o) OPENVINO_VERSION=${OPTARG};;
 esac
 done
 
-OPENVINO_VERSION=${OPENVINO_VERSION:=2020.2}
-export INTEL_OPENVINO_DIR=/data/dldt/openvino_${OPENVINO_VERSION}
-git clone https://github.com/opencv/dldt.git ${INTEL_OPENVINO_DIR}
+OPENVINO_VERSION=${OPENVINO_VERSION:=2020.4}
+export INTEL_OPENVINO_DIR=/data/openvino/openvino_${OPENVINO_VERSION}.287
```
Comment (HectorSVC) [inline tools/ci_build/github/linux/docker/scripts/install_openvino.sh:11]: >287 [](start = 70, length = 3)

should we make the minor version as a parameter also? so we don't need to always change this file.

File: tools/ci_build/github/linux/docker/Dockerfile.ubuntu_openvino [hunk 3]
```diff
@@ -17,8 +17,8 @@ RUN /tmp/scripts/install_openvino.sh -o ${OPENVINO_VERSION} && \
 
 WORKDIR /root
 
-ENV INTEL_OPENVINO_DIR /data/dldt/openvino_${OPENVINO_VERSION}
-ENV LD_LIBRARY_PATH $INTEL_OPENVINO_DIR/deployment_tools/inference_engine/lib/intel64:$INTEL_OPENVINO_DIR/deployment_tools/:$INTEL_OPENVINO_DIR/deployment_tools/ngraph/lib:$INTEL_OPENVINO_DIR/deployment_tools/inference_engine/external/tbb/lib:/usr/local/openblas/lib:$LD_LIBRARY_PATH
+ENV INTEL_OPENVINO_DIR /data/openvino/openvino_${OPENVINO_VERSION}.287
```
Comment (HectorSVC) [inline tools/ci_build/github/linux/docker/Dockerfile.ubuntu_openvino:20]: >287 [](start = 67, length = 3)

should we make the minor version as a parameter also? so we don't need to always change this file.

**Comment by HectorSVC** [comment]
don't know why the files build_yocto.sh & yocto_build_toolchain.sh were included. Please revert the changes on these 2 files.

**Comment by suryasidd** [comment]
> don't know why the files build_yocto.sh & yocto_build_toolchain.sh were included. Please revert the changes on these 2 files.

reverted.

**Comment by HectorSVC** [comment]
> There's a test failure (result diff), please take a look.
> 2020-08-18T07:09:50.0264924Z 1: [ RUN ] ModelTests/ModelTest.Run/719
> 2020-08-18T07:09:50.0313166Z 1: Warning: ImageScaler was a removed experimental ops. In the future, we may directly reject this operator. Please update your model as soon as possible.
> 2020-08-18T07:09:50.0314320Z 1: Warning: Crop was a removed experimental ops. In the future, we may directly reject this operator. Please update your model as soon as possible.
> 2020-08-18T07:09:50.0315301Z 1: Warning: Crop was a removed experimental ops. In the future, we may directly reject this operator. Please update your model as soon as possible.
> 2020-08-18T07:09:50.0316198Z 1: Warning: Crop was a removed experimental ops. In the future, we may directly reject this operator. Please update your model as soon as possible.
> 2020-08-18T07:09:50.0317086Z 1: Warning: Crop was a removed experimental ops. In the future, we may directly reject this operator. Please update your model as soon as possible.
> 2020-08-18T07:09:50.0321897Z 1: Warning: Crop was a removed experimental ops. In the future, we may directly reject this operator. Please update your model as soon as possible.
> 2020-08-18T07:09:50.0322998Z 1: Warning: Crop was a removed experimental ops. In the future, we may directly reject this operator. Please update your model as soon as possible.
> 2020-08-18T07:09:50.0323965Z 1: Warning: Crop was a removed experimental ops. In the future, we may directly reject this operator. Please update your model as soon as possible.
> 2020-08-18T07:09:50.0324961Z 1: Warning: Affine was a removed experimental ops. In the future, we may directly reject this operator. Please update your model as soon as possible.
> 2020-08-18T07:11:22.4653749Z 1: /onnxruntime_src/onnxruntime/test/providers/cpu/model_tests.cc:556: Failure
> 2020-08-18T07:11:22.4655738Z 1: Expected equality of these values:
> 2020-08-18T07:11:22.4657791Z 1: COMPARE_RESULT::SUCCESS
> 2020-08-18T07:11:22.4659587Z 1: Which is: 4-byte object <00-00 00-00>
> 2020-08-18T07:11:22.4660015Z 1: ret.first
> 2020-08-18T07:11:22.4660586Z 1: Which is: 4-byte object <01-00 00-00>
> 2020-08-18T07:11:22.4661154Z 1: expected 132 (43040000), got 142.494 (430e7e7a), diff: 10.494, tol=10.0013 idx=851956. 1 of 1555200 differ
> 2020-08-18T07:11:22.5404617Z 1: [ FAILED ] ModelTests/ModelTest.Run/719, where GetParam() = "openvino_../models/opset7/coreml_FNS-Candy_ImageNet/winmlperf_coreml_FNS-Candy_prerelease.onnx" (92514 ms)

We should ignore this model.

File: onnxruntime/core/providers/openvino/ov_versions/capabilities.h (hunk 1 above)
Comment (smkarlap) [reply onnxruntime/core/providers/openvino/ov_versions/capabilities.h:14]: Yes, using inheritance is certainly a cleaner solution. Having fewer checks/restrictions with newer OpenVINO versions is generally true with only GetCapability code. Refactoring this is something that we are planning to do for the next release. 

For other areas of code (Compile and Infer functions) it gets a bit more complex as code working with newer OpenVINO versions does some things differently due to newer API features available. With this area, we are still considering on how to refactor the logic more cleanly.

File: dockerfiles/Dockerfile.openvino [hunk 4]
```diff
@@ -6,19 +6,16 @@
 FROM ubuntu:18.04
```
Comment (jywu-msft) [inline dockerfiles/Dockerfile.openvino:6]: Thanks. I tested this new Dockerfile and it seems to work well. Glad we don't have to manually download the software anymore.

File: tools/ci_build/github/linux/docker/Dockerfile.ubuntu_openvino (hunk 3 above)
Comment (jywu-msft) [reply tools/ci_build/github/linux/docker/Dockerfile.ubuntu_openvino:20]: yes, seems like the version should include the minor version.

**Pull request merged**